pkg/controller/plan/adapter/openstack/builder.go

```diff
@@ -61,17 +210,203 @@ func (r *Builder) VirtualMachine(vmRef ref.Ref, object *cnv.VirtualMachineSpec,
 	return
 }
 
+func (r *Builder) mapFirmware(vm *model.Workload, object *cnv.VirtualMachineSpec) {
+	var firmwareType string
+	var bootloader *cnv.Bootloader
+	if imageFirmwareType, ok := vm.Image.Properties[FirmwareType]; ok {
+		firmwareType = imageFirmwareType.(string)
+	} else {
+		for _, volume := range vm.Volumes {
+			if volume.Bootable == "true" {
+				if volumeFirmwareType, ok := volume.VolumeImageMetadata[FirmwareType]; ok {
+					firmwareType = volumeFirmwareType
+				}
+			}
+		}
+	}
+	switch firmwareType {
+	case EFI:
+		// We disable secure boot even if it was enabled on the source because the guest OS won't
+		// be able to boot without getting the NVRAM data. So we start the VM without secure boot
+		// to ease the procedure users need to do in order to make the guest OS to boot.
+		secureBootEnabled := false
+		bootloader = &cnv.Bootloader{
+			EFI: &cnv.EFI{
+				SecureBoot: &secureBootEnabled,
+			}}
+	default:
+		bootloader = &cnv.Bootloader{BIOS: &cnv.BIOS{}}
+	}
+	features := &cnv.Features{}
+	firmware := &cnv.Firmware{}
+	firmware.Bootloader = bootloader
+	object.Template.Spec.Domain.Features = features
+	object.Template.Spec.Domain.Firmware = firmware
+}
+
+func (r *Builder) mapVideo(vm *model.Workload, object *cnv.VirtualMachineSpec) {
+	videoModel := DefaultProperties[VideoModel]
+	if imageVideoModel, ok := vm.Image.Properties[VideoModel]; ok {
+		videoModel = imageVideoModel.(string)
+	}
+	switch videoModel {
+	case None:
+	default:
+		autoAttachGraphicsDevice := false
+		object.Template.Spec.Domain.Devices.AutoattachGraphicsDevice = &autoAttachGraphicsDevice
+	}
+}
+
+func (r *Builder) mapInput(vm *model.Workload, object *cnv.VirtualMachineSpec) {
+	if _, ok := vm.Image.Properties[PointerModel]; ok {
+		tablet := cnv.Input{
+			Type: Tablet,
+			Name: Tablet,
+			Bus:  USB,
+		}
+		object.Template.Spec.Domain.Devices.Inputs = []cnv.Input{tablet}
+	}
+}
+
+func (r *Builder) mapCPUMem(vm *model.Workload, object *cnv.VirtualMachineSpec) {
+	// Set MachineType
+	machineType := DefaultProperties[MachineType]
+	if imageMachineType, ok := vm.Image.Properties[MachineType]; ok {
+		machineType = imageMachineType.(string)
+	} else {
+		for _, volume := range vm.Volumes {
+			if volume.Bootable == "true" {
+				if volumeMachineType, ok := volume.VolumeImageMetadata[MachineType]; ok {
+					machineType = volumeMachineType
+				}
+			}
+		}
+	}
+	object.Template.Spec.Domain.Machine = &cnv.Machine{Type: machineType}
```

Comment:
note that the CI fails on:
spec.template.spec.domain.machine.type is not supported: pc (allowed values: [q35* pc-q35*])]

Reply:
I have changed this to always use Q35 as it's the only supported machine type.